src/weighted_list.h

```diff
@@ -0,0 +1,107 @@
+#ifndef WEIGHTED_LIST_H
+#define WEIGHTED_LIST_H
+
+#include "rng.h"
+#include <vector>
+
+template <typename T> struct weighted_obect {
+    T obj;
+    int weight;
+};
+
+template <typename T> struct weighted_list {
+    weighted_list() : total_weight(0) { };
+
+    /**
+     * This will add a new object to the weighted list.
+     * @param obj The object that will be added to the list.
+     * @param weight The weight of the object.
+     */
+    void add_item(T obj, int weight) {
+        if(weight >= 0) {
+            weighted_obect<T> new_weighted = { obj, weight };
+            objects.push_back(new_weighted);
+            total_weight += weight;
+        }
+    }
+
+    /**
+     * This will check to see if the given object is already in the weighted
+       list. If it is, we update its weight with the new weight value. If it
+       is not, we add it normally.
+     * @param obj The object that will be updated or added to the list.
+     * @param weight The new weight of the object.
+     */
+    void add_or_replace_item(T obj, int weight) {
+        if(weight >= 0) {
+            for(auto &itr : objects) {
+                if(itr.obj == obj) {
+                    total_weight += (weight - itr.weight);
+                    itr.weight = weight;
+                    return;
+                }
+            }
+
+            // if not found, add to end of list
+            add_item(obj, weight);
+        }
+    }
+
+    /**
+     * This will call the given callback function once for every object in the weighted list.
+     * @param func The callback function.
+     */
+    void apply(void (*func)(T &obj)) {
+        for(auto &itr : objects) {
+            func(itr.obj);
+        }
+    }
+
+    /**
+     * This will return a pointer to an object from the list randomly selected
+     * and biased by weight. If the weighted list is empty or all items in it
+     * have a weight of zero, it will return a NULL pointer.
```

Comment:
The multiplying was my suggesting to create a standard scale at runtime, one needs to know the factor anyway when picking an entry. My suggesting is basically: load data from json, convert weight (from json in any format) to a value in a standard scale, add the object with that normalized weight.

----

For negative values you also need to change the summation of the total_weight (add `abs(weight)` or so) and the code in `pick_entry`. Your probably better of with a separate implementation of it, which would also be clearer and easier to understand: different classes - different behavior.

Making the weight type templated seems overkill and primarily adds different data, not different logic. You would need to add specializations for different weight types that 1. sum the total weight differently, 2. randomly pick differently. And there is no reason why this should be different based on the *type* of the weight.

For example: You can have negative weight (with special meaning) using `int`, but also when using `float`. You can also have only positive weight value (with no special meaning, negative values would be errors) and use `float` or `int`. In other words: the type does not tell you anything.